### PR TITLE
In-place after view in no_grad (issue #26546)

### PR DESCRIPTION
**Pull request by sublee**
Issue: #26546

In `torch.no_grad()` mode, an in-place operation on a view tensor (`t()`, `view()`, `squeeze()`, etc.) attaches grad_fn `AsStridedBackward` to the tensor.

For example, `y` in the below code should not have any `grad_fn` because it has passed through `torch.no_grad()`:

```python
x = torch.ones(1, requires_grad=True)
with torch.no_grad():
    y = x.t()
    y.add_(0)
```

But it has `grad_fn`:

```python
>>> y
tensor([1.], grad_fn=<AsStridedBackward>)
```

This behavior may cause a bug difficult to understand when we compose InstanceNorm (it makes a view tensor internally) and in-place ReLU in a checkpoint:

```python
x = torch.ones(3, 2, 1, requires_grad=True)
model = nn.Sequential(
    nn.InstanceNorm1d(2),
    nn.ReLU(inplace=True),
)
y = checkpoint(model, x)
y.norm().backward()

# x.grad should be tensor(...). But it is still None because it didn't receive any gradients.
assert x.grad is not None  # FAILS!

# y.grad_fn should be <CheckpointFunctionBackward>. But it is <AsStridedBackward> actually.
assert y.grad_fn.__class__ is CheckpointFunction._backward_cls  # FAILS!
```

To access `Tensor.grad_fn` in Python triggers `Variable::grad_fn()` in [autograd/variable.cpp](https://github.com/pytorch/pytorch/blob/v1.2.0/torch/csrc/autograd/variable.cpp#L135-L162). This method always returns `<AsStridedBackward>` if the tensor is a view with in-place mutation regardless of the grad mode it has passed through. So I've tried to make the tensor as not a view if the view was created in `no_grad` mode.

To suppress `DifferentiableViewMeta` in `no_grad` mode seems to fix those problems. I just made `as_view()` check `GradMode::is_enabled()` before calling `make_view_variable()`.

```diff
  inline Tensor as_view(const Tensor & base, Tensor tensor, bool is_differentiable = true) {
    auto base_var = Variable(base);
    if (base_var.is_view()) {
      base_var = base_var.base();
    }

+   if (!GradMode::is_enabled()) {
+     is_differentiable = false;
+   }

    return make_variable_view(std::move(base_var), std::move(tensor), is_differentiable);
  }
```

Please review this solution and the following code.

### CODE REVIEW

**Comment by sublee** [comment]
If the output of a custom autograd function is a view tensor mutated in-place, its `backward()` will never be executed. I think it's a serious problem so I reproduced it as a test case. This PR fixes also this problem.

```python
def test_inplace_view_in_autograd_function(self):
    # See https://github.com/pytorch/pytorch/issues/26546
    class F(torch.autograd.Function):
        @staticmethod
        def forward(ctx, x):
            y = x.t()
            y.add_(0)
            return y

        @staticmethod
        def backward(ctx, g):
            return g.fill_(42)

    x = torch.rand(3, requires_grad=True)
    y = F.apply(x)
    y.sum().backward()

    # y.grad_fn should be <FBackward>, but it is <AsStridedBackward> actually.
    assert y.grad_fn.__class__ is F._backward_cls

    # x.grad.mean() should be 42, but it is 1 actually.
    assert x.grad.mean().item() == 42
```

**Comment by albanD** [comment]
Hi,

Sorry for the late reply.
This is definitely something we want to fix, but there are a few other edge case that we should think of to get the best solution.
I'll get back to you tomorrow !

**Comment by albanD** [comment]
Hi,

I agree that the behavior for `autograd.Function` is clearly wrong at the moment.

The behavior within the `torch.no_grad()` block though is less clear. In particular if you do:
```python
x = torch.ones(10, requires_grad=True)
with torch.no_grad():
    y = x[1:]
y.add_(0)
```
Your change will prevent gradients from flowing back all the way to `x`. While `y` is just a subset of `x` and so you might expect gradients to flow back.

In particular current users might be relying on this behavior and so this PR is introducing breaking changes.


We would be happy to change the behavior for `Function` but the impact of the current diff are quite wide and hard to identify.
In particular, changing the `autograd.Function` to make sure that if an output is a view, it is either a view of an input (and the current logic handles that) or a view to a temporary Tensor that does not require gradients (and new logic should mark it as not being a view anymore).
Some other edge case may appear (what if you return two outputs and one is a view of the other?) and should be clearly identified and either handled or explicitly forbidden with an error.

**Review by albanD — CHANGES_REQUESTED** [review]
Hi,

I agree that the behavior for autograd.Function is clearly wrong at the moment.

The behavior within the torch.no_grad() block though is less clear. In particular if you do:

x = torch.ones(10, requires_grad=True)
with torch.no_grad():
    y = x[1:]
y.add_(0)
Your change will prevent gradients from flowing back all the way to x. While y is just a subset of x and so you might expect gradients to flow back.

In particular current users might be relying on this behavior and so this PR is introducing breaking changes.

We would be happy to change the behavior for Function but the impact of the current diff are quite wide and hard to identify.
In particular, changing the autograd.Function to make sure that if an output is a view, it is either a view of an input (and the current logic handles that) or a view to a temporary Tensor that does not require gradients (and new logic should mark it as not being a view anymore).
Some other edge case may appear (what if you return two outputs and one is a view of the other?) and should be clearly identified and either handled or explicitly forbidden with an error.

**Comment by sublee** [comment]
Hi @albanD, thank you for considering this PR.

I agree that this PR may cause breaking changes to users. I'll try to spend some time to fix `torch.autograd.Function` rather than `view_as()`.

But I don't understand the expected gradient flow of your example code. `x` doesn't seem to should receive gradients through `y` for me:

```python
x = torch.ones(10, requires_grad=True)
with torch.no_grad():
    y = x[1:]
y.add_(0)
```

`torch.no_grad()` is an explicit API to disable automatic dependency for flowing gradients. So I thought that when users use explicit `torch.no_grad()` over a view function, they would expect the view tensor is detached from the base tensor. I guessed both the below code represent the same intention:

```python
with torch.no_grad():
    b = a.view(a.size())
```

```python
b = a.detach()
```

Would you explain why should `x` receive gradients through `y` in your example code?

**Comment by albanD** [comment]
It is a fairly philosophical question but you could argue that the differentiability is a property of a given entry of the Tensor.
And so whatever you do with the entries in `x` it should participate in it's gradient computation.
But under that assumption, the meaning of `torch.no_grad()` is not so clear anymore.

I am currently working on discussing with all the people involved to try and find the best solution for this.

**Comment by sublee** [comment]
@albanD Thanks to discussing deeply this issue. I'm looking forward to the best solution and a more clear definition of `torch.no_grad()` or in-place view. The conclusion of the people would help me to catch the philosophical background of this issue.

**Comment by sublee** [comment]
@albanD seems to have involved in this issue. I close this pull request.
